File: internal/approval/approval.go
package approval

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"database/sql"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"github.com/mohamedmasoud3030-tech/mosaid/internal/audit"
	"time"
)

type Manager struct {
	DB    *sql.DB
	Audit audit.Logger
}
type Request struct {
	ID                       string
	Token                    string
	UserID                   int64
	Tool, ArgsHash, Resource string
	Expires                  time.Time
}

func (m Manager) Create(ctx context.Context, user int64, tool, argsHash, resource string, ttl time.Duration) (Request, error) {
	b := make([]byte, 24)
	if _, e := rand.Read(b); e != nil {
		return Request{}, e
	}
	tok := base64.RawURLEncoding.EncodeToString(b)
	th := sha256.Sum256([]byte(tok))
	idb := sha256.Sum256(append([]byte("id:"), b...))
	r := Request{ID: hex.EncodeToString(idb[:12]), Token: tok, UserID: user, Tool: tool, ArgsHash: argsHash, Resource: resource, Expires: time.Now().UTC().Add(ttl)}
	_, e := m.DB.ExecContext(ctx, `INSERT INTO approval_requests(id,token_hash,user_id,tool_name,args_hash,resource,expires_at,state,created_at) VALUES(?,?,?,?,?,?,?,'pending',?)`, r.ID, hex.EncodeToString(th[:]), user, tool, argsHash, resource, r.Expires.Format(time.RFC3339Nano), time.Now().UTC().Format(time.RFC3339Nano))
	if e == nil {
		_, _ = m.Audit.Append(ctx, audit.Entry{Kind: "approval_requested", UserID: user, Resource: resource, Decision: "pending"})
	}
	return r, e
}
func (m Manager) ResolveToken(ctx context.Context, token string, user int64, decision string) error {
	th := sha256.Sum256([]byte(token))
	var tool, argsHash, resource string
	var storedUser int64
	if err := m.DB.QueryRowContext(ctx, `SELECT user_id,tool_name,args_hash,resource FROM approval_requests WHERE token_hash=?`, hex.EncodeToString(th[:])).Scan(&storedUser, &tool, &argsHash, &resource); err != nil {
		return errors.New("approval not found")
	}
	if storedUser != user {
		return errors.New("approval binding mismatch")
	}
	return m.Resolve(ctx, token, user, tool, argsHash, resource, decision)
}
func (m Manager) Authorize(ctx context.Context, token string, user int64, tool, argsHash, resource string) error {
	th := sha256.Sum256([]byte(token))
	tx, e := m.DB.BeginTx(ctx, nil)
	if e != nil {
		return e
	}
	defer tx.Rollback()
	var id, t, a, r, state string
	var u int64
	e = tx.QueryRowContext(ctx, `SELECT id,user_id,tool_name,args_hash,resource,state FROM approval_requests WHERE token_hash=?`, hex.EncodeToString(th[:])).Scan(&id, &u, &t, &a, &r, &state)
	if e != nil {
		return errors.New("approval not found")
	}
	if state != "approved" || u != user || t != tool || a != argsHash || r != resource {
		return errors.New("approval not authorized")
	}
	if _, e = tx.ExecContext(ctx, `INSERT INTO approval_uses(approval_id,used_at) VALUES(?,?)`, id, time.Now().UTC().Format(time.RFC3339Nano)); e != nil {
		return errors.New("approval already used")
	}
	return tx.Commit()
}
func (m Manager) Resolve(ctx context.Context, token string, user int64, tool, argsHash, resource, decision string) error {
	if decision != "approved" && decision != "denied" {
		return errors.New("invalid decision")
	}
	th := sha256.Sum256([]byte(token))
	tx, e := m.DB.BeginTx(ctx, nil)
	if e != nil {
		return e
	}
	defer tx.Rollback()
	var id, t, a, r, exp, state string
	var u int64
	e = tx.QueryRowContext(ctx, `SELECT id,user_id,tool_name,args_hash,resource,expires_at,state FROM approval_requests WHERE token_hash=?`, hex.EncodeToString(th[:])).Scan(&id, &u, &t, &a, &r, &exp, &state)
	if e != nil {
		return errors.New("approval not found")
	}
	if state != "pending" {
		return errors.New("approval already resolved")
	}
	et, _ := time.Parse(time.RFC3339Nano, exp)
	if time.Now().UTC().After(et) {
		_, _ = tx.ExecContext(ctx, `UPDATE approval_requests SET state='expired' WHERE id=?`, id)
		_ = tx.Commit()
		return errors.New("approval expired")
	}
	if u != user || t != tool || a != argsHash || r != resource {
		return errors.New("approval binding mismatch")
	}
	res, e := tx.ExecContext(ctx, `UPDATE approval_requests SET state=?,resolved_at=? WHERE id=? AND state='pending'`, decision, time.Now().UTC().Format(time.RFC3339Nano), id)
	if e != nil {
		return e
	}
	n, _ := res.RowsAffected()
	if n != 1 {
		return errors.New("approval replay")
	}
	if e = tx.Commit(); e != nil {
		return e
	}
	_, _ = m.Audit.Append(ctx, audit.Entry{Kind: "approval_resolved", UserID: user, Resource: resource, Decision: decision})
	return nil
}

File: mosaid-cloud/lib/core/approval.ts
import { createHash, randomUUID } from "crypto";
import type { SupabaseClient } from "@supabase/supabase-js";
import { Audit } from "./audit";

/**
 * مدير الموافقات — ينقل مبادئ internal/approval في Mosaid:
 *  - نُخزّن فقط hash للتوكن (لا توكن عادي)
 *  - الطلب مرتبط بالأداة + argsHash + المورد
 *  - TTL (انتهاء الصلاحية)
 *  - استخدام مرة واحدة (approval_uses) ضد إعادة الاستخدام
 *  - سجل تدقيق عند الطلب/القرار
 */

const TTL_MS = 10 * 60 * 1000; // 10 دقائق

function sha256(s: string): string {
  return createHash("sha256").update(s).digest("hex");
}

export async function createApproval(
  supabase: SupabaseClient,
  req: { tool: string; argsHash: string; resource: string }
): Promise<{ id: string; token: string; expires: string }> {
  const token = randomUUID();
  const tokenHash = sha256(token);
  const expires = new Date(Date.now() + TTL_MS).toISOString();

  const { data, error } = await supabase
    .from("approval_requests")
    .insert({
      token_hash: tokenHash,
      tool_name: req.tool,
      args_hash: req.argsHash,
      resource: req.resource,
      state: "pending",
      expires_at: expires,
    })
    .select()
    .single();
  if (error) throw new Error("approval create failed: " + (error.message || "unknown"));

  await Audit(supabase, {
    kind: "approval_requested",
    resource: req.resource,
    decision: "pending",
    detail: { tool: req.tool },
  });

  return { id: (data as { id: string }).id, token, expires };
}

export async function resolveApproval(
  supabase: SupabaseClient,
  id: string,
  decision: "approved" | "denied"
): Promise<void> {
  const { data, error } = await supabase
    .from("approval_requests")
    .select("*")
    .eq("id", id)
    .single();
  if (error) throw new Error("approval not found");
  const row = data as { state: string; tool_name: string; resource: string };

  if (row.state !== "pending") throw new Error(`approval already ${row.state}`);
  if (!decision || (decision !== "approved" && decision !== "denied")) throw new Error("invalid decision");

  const { error: upErr } = await supabase
    .from("approval_requests")
    .update({ state: decision, resolved_at: new Date().toISOString() })
    .eq("id", id)
    .eq("state", "pending");
  if (upErr) throw new Error("approval resolve failed: " + upErr.message);

  await Audit(supabase, {
    kind: "approval_resolved",
    resource: row.resource,
    decision,
    detail: { tool: row.tool_name },
  });
}

/**
 * تنفيذ الأثر الفعلي بعد الموافقة، واستهلاك الطلب (مرة واحدة).
 */
export async function authorizeAndApply(
  supabase: SupabaseClient,
  id: string,
  apply: (payload: Record<string, unknown>) => Promise<void>
): Promise<void> {
  const { data, error } = await supabase.from("approval_requests").select("*").eq("id", id).single();
  if (error) throw new Error("approval not found");
  const row = data as {
    id: string;
    tool_name: string;
    resource: string;
    state: string;
    args_hash: string;
  };
  if (row.state !== "approved") throw new Error(`approval not approved (state=${row.state})`);

  // استخدام مرة واحدة
  const { data: uses, error: uErr } = await supabase
    .from("approval_uses")
    .select("id")
    .eq("approval_id", id);
  if (uErr) throw new Error("use check failed");
  if ((uses ?? []).length > 0) throw new Error("approval already used");

  // نبني البايلود من args_hash؟ نتوقف عند هذا لا: نستدعي effect بواسطة args_hash
  // فيتطلب من المتصل تمرير القيمة. نُمرّره إذا كان في الطلب.
  await apply({ tool: row.tool_name, resource: row.resource, argsHash: row.args_hash });

  await supabase.from("approval_uses").insert({ approval_id: id });

  await Audit(supabase, {
    kind: "approval_applied",
    resource: row.resource,
    decision: "applied",
    detail: { tool: row.tool_name },
  });
}
